[lldb][RPC] Upstream RPC server interface emitters

chelcassanova opened the pull request:
This commit upstreams the LLDB RPC server interface emitters. These emitters generate the server-side API interfaces for RPC, which communicate directly with liblldb itself out of process using the SB API.

https://discourse.llvm.org/t/rfc-upstreaming-lldb-rpc/85804

chelcassanova committed:
[lldb[RPC] Upstream RPC server interface emitters

This commit upstreams the LLDB RPC server interface emitters. These
emitters generate the server-side API interfaces for RPC, which
communicate directly with liblldb itself out of process using the SB
API.

https://discourse.llvm.org/t/rfc-upstreaming-lldb-rpc/85804

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/server/RPCServerSourceEmitter.cpp:
```diff
@@ -0,0 +1,592 @@
+void RPCServerSourceEmitter::EmitMethodCallAndEncode(const Method &method) {
+  // FIXME: The hand-written lldb-rpc-server currently doesn't emit destructors
+  // for LocalObjectRefs... even if the type is an ObjectRef. What should we do
+  // here?
```
What's the impact of this and, to repeat the comment, what should we do here?

Is this a nice to have or a memory leak waiting to happen if someone uses a LocalObjectRef in the right/wrong way?

(whatever a LocalObjectRef is :) )

bulbazord replied:
I think this `FIXME` can be removed. There _should_ be no impact because a LocalObjectRef's only member is a shared pointer to an ObjectRef (which does have a proper destructor).

Future PRs should have the proper definitions for ObjectRef and LocalObjectRef, but effectively every SB class inherits from one of them. ObjectRef holds onto a connection, a class ID (so you know what kind of object it is), and an object ID (so you know which object it refers to).

chelcassanova replied:
And to add on to this, `ObjectRefs` are held on to by the server (they're usually used to store larger amounts of information about an RPC object and its connection`) while `LocalObjectRef`s are used locally for smaller classes (like Lists and Specs if I'm not mistaken).

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/server/RPCServerSourceEmitter.cpp:
```diff
@@ -0,0 +1,592 @@
+  if (!method.IsCtor) {
+    // We need to determine what the appropriate return type is. Here is the
+    // strategy:
+    // 1.) `SBFoo` -> `SBFoo &&`
```
So if have SBFoo you return rvalue reference to SBFoo?

I don't claim to understand `&&` in normal circumstances either, so maybe I am just confused.

bulbazord replied:
This comment is a bit misleading (and I wrote it, so you can blame me :p).

The code we generated here would look like this:
```
SBFoo &&__result = this_ptr->GetTheFoo();
RPCServerObjectEncoder(response, rpc_common::RPCPacket::ValueType::ReturnValue, std::move(__result));
return true;
```

`this_ptr->GetTheFoo` would create a temporary `SBFoo` value and `__result` is a reference (rvalue ref) to that temporary value. This extends the lifetime of that temporary `SBFoo`.

On the next line, we'll "encode" that result value into the `response` stream (which is going to get turned into a packet). This line does a lot of work that you don't have access to yet, but effectively it will:
1. Put the SBFoo into an "ObjectInstance" map. The client and server speak in terms of "object IDs" instead of moving information about an object in their protocol packets. The server maintains a mapping of (Object ID -> Object Instance), and when we encode the SBFoo return value, the instance gets assigned an Object ID and stuck in the map.
2. Encode the return value's object ID into the "response" packet and indicate that it is a return value from this method.
3. Returning "true" indicates that the RPC call succeeded on the server-side.

DavidSpickett replied:
Understood, thanks.

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/server/RPCServerSourceEmitter.cpp:
```diff
@@ -0,0 +1,592 @@
+        EmitLine(
+            "// Intentionally inserting a compiler error. lldb-rpc-gen "
+            "was unable to determine how large the return buffer should be.");
+        EmitLine("ThisShouldNotCompile");
```
`#error <something vaguely useful>` would be a bit more standard and provide some help to debug this.

DavidSpickett replied:
This still needs doing.

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/server/RPCServerSourceEmitter.h:
```diff
@@ -0,0 +1,80 @@
+using namespace clang;
+
+namespace lldb_rpc_gen {
+class RPCServerSourceEmitter : public FileEmitter {
```
Please add a top level description to each of these new classes. Even if it is just "Emits the .cpp files for bla", at least then I know that it's doing headers or code or both maybe.

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/server/RPCServerSourceEmitter.cpp:
```diff
@@ -0,0 +1,592 @@
+      QualType UnqualifiedUnderlyingSBClass =
+          UnderlyingSBClass.getUnqualifiedType();
+
+      // FIXME: Replace this bespoke logic with a nice function in RPCCommon.
```
Does RPCCommon actually exist or is that also a future thing?

Either way doesn't sound critical, just wondering.

bulbazord replied:
It's in a separate PR that Chelsea created. There's a circular dependency between this PR and that PR unfortunately.

DavidSpickett reviewed (commented):
Looks OK in isolation, a bit like looking at a TableGen backend, easier to judge the output than the emitters.

Please look over the FIXMEs and see if anything can be done right now, I highlighted one that is potentially a problem but the rest don't seem important. See if you agree.

Also please remove any that are no longer relevant.

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/server/RPCServerSourceEmitter.cpp:
```diff
@@ -0,0 +1,592 @@
+  // First, we consider `const char *`, `const char **`. They have special
+  // server-side types.
+  if (TypeIsConstCharPtr(ParamType)) {
+    EmitLine("rpc_common::ConstCharPointer " + ParamName + ";");
```
Ok, I saw a note about this somewhere and thought it was literally a typedef not a wrapper object.

DavidSpickett commented on lldb/tools/lldb-rpc/lldb-rpc-gen/server/RPCServerSourceEmitter.cpp:
```diff
@@ -0,0 +1,592 @@
+// For methods with pointer return types, it's important that we know how big
+// the type of the pointee is. We must correctly size a buffer (in the form of a
+// Bytes object) before we can actually use it.
+static const std::map<llvm::StringRef, size_t> MethodsWithPointerReturnTypes = {
```
So even if you did `sizeof(pointee type)`, it could still be pointing to more than one `uint8_t`?

DavidSpickett replied:
If only someone was working on bounds annotations... :)

Understood.

bulbazord commented:
> Pushed again to address David's comments and remove most of the FIXMEs. There's one FIXME I've kept in for now:
> 
> ```
> // FIXME: SB class server references are stored as non-const references so
> // that we can actually change them as needed. If a parameter is marked
> // const, we will fail to compile because we cannot make an
> // SBFooServerReference from a `const SBFoo &`.
> // To work around this issue, we'll apply a `const_cast` if needed so we
> // can continue to generate callbacks for now, but we really should
> // rethink the way we store object IDs server-side to support
> // const-qualified parameters.
> ```
> 
> @bulbazord This is referring to how we emit storage for const SB parameters in callback functions. Currently it looks like it only gets applied to one function that gets generated. Any thoughts here?

Given it only applies to one function, it's probably okay to keep it around for now. The fact that we need a `const_cast` at all is concerning but I think we've done a reasonably alright job at explaining why it exists (though I am a partial author, so maybe my perspective is biased).

DavidSpickett reviewed (commented):
A couple of comments to be addressed but otherwise I have no problems with this.

chelcassanova commented:
I added some small shell tests to this patch that check the output of the emitter mainly for sanity checking. @bulbazord I think having basic tests like what I added would be good for the server-side emitter, but if there's anything other thing you can think of test here please let me know!

chelcassanova commented:
@DavidSpickett I think I've address any previous comments to this point and that this patch is ready to land (contingent on the rpc-gen tool itself landing first). If possible, could you give this another once over?

DavidSpickett commented on lldb/test/Shell/RPC/Generator/Tests/Server/CheckBasicIncludesEmit.test:
```diff
@@ -0,0 +1,15 @@
+# Disabling until the lldb-rpc-gen tool lands.
+UNSUPPORTED: system-windows, system-linux, system-darwin
```
`UNSUPPORTED: *` should disable it everywhere.

Saves someone on FreeBSD getting an unpleasant surprise.

DavidSpickett replied:
Do that for the rest of the tests as well.

chelcassanova replied:
The unsupported can probably be removed actually, these tests still need to test after the tool lands however.

DavidSpickett reviewed (approved):
LGTM with the UNSUPPORTED changed to apply to any system.
